### ISSUE TITLE
Simplify Configs

### ISSUE THREAD

**Issue by mivanit** [understanding-search/maze-transformer#10]
tl;dr: model/training/dataset configuration is a mess and needs to be reworked

Currently, we have 3 different configurations:
- Model configurations, which deal with architecture of the transformer nets and create configs which inherit from hugginface model configs
- Training configurations, which handle training hyperparams
- Dataset config, which tells us how the dataset was generated and what size mazes it contains

the problem we've run into so far is that its easy to forget to adjust the vocab size in the model config, which should be inherited somehow from the dataset config. Moving forward, we also want to consider that we want to account for:
- non-lattice mazes (arbitrary graphs)
- higher-dimensional mazes
- different tokenization methods
  - this includes tokenizing the x,y coord of a cell in the maze separately, rather than having a unique token for each cell
- other maze properties, such as target/negative reward regions, labelling the maze cells with additional data, etc
- finetuning models on a different dataset from the original dataset

my naive instinct is that the structure of configs should be something like the following:

```python
@dataclass
class ModelConfig:
  train_config: TrainConfig|None
  dataset_config: DatasetConfig|None
  # (...) other properties of the architecture

@dataclass
class BaseModel:
  model_config: ModelConfig
  training_log: list|None # somewhere to log arbitrary data about the training process in a semi-human readable format
```

things like vocab size could be properties made to inherit from the `DatasetConfig`, and thus raise exceptions if the required config is not provided, rather than silently give the wrong values
